ocaml/xapi/xapi_vtpm.ml

```diff
@@ -12,19 +12,43 @@
    GNU Lesser General Public License for more details.
  *)
 
+let assert_no_vtpm_associated ~__context vm =
+  match Db.VM.get_VTPMs ~__context ~self:vm with
+  | [] ->
+      ()
+  | vtpms ->
+      let uuids =
+        List.map (fun v -> Db.VTPM.get_uuid ~__context ~self:v) vtpms
+      in
+      raise Api_errors.(Server_error (vtpm_max_amount_reached, uuids))
+
+let assert_vm_is_halted ~__context vm =
+  let attachable = [`Halted] in
+  match Db.VM.get_power_state ~__context ~self:vm with
+  | state when List.exists (( = ) state) attachable ->
```

Comment:
using ``Halted` directly in the match would be shorter?

Reply:
Also would be good to have an allowed operation on the VM object that tells you when you can or cannot create vtpms, and an allowed operation on the vtpm object to tell you when you can destroy it.
Then this can be updated (and hopefully this assertion code shared). I think the usual way is to put it into some lifecycle helper module and call the assertion in a strict mode when performing the API call.

Reply:
> using ``Halted` directly in the match would be shorter?
You're right, I was thinking of adding more that one state there and avoid duplication when generating the error message

I'm wary of adding vtpm creation / deletion to the allowed operations of the VM, as the operation is on the device, not the domain.

Having allowed operations on the vtpm would more sense if there were several operations there (I was thinking about attaching and detaching). Since it's a single operation, I'm not sure this would be of much benefit right now to have this.